modules/040-control-plane-manager/hooks/generate_admission_webhook_cert.go

```diff
@@ -0,0 +1,113 @@
+/*
+Copyright 2021 Flant JSC
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package hooks
+
+import (
+	"fmt"
+	"time"
+
+	"github.com/cloudflare/cfssl/csr"
+	"github.com/flant/addon-operator/pkg/module_manager/go_hook"
+	"github.com/flant/addon-operator/sdk"
+	"github.com/flant/shell-operator/pkg/kube_events_manager/types"
+	v1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+
+	"github.com/deckhouse/deckhouse/go_lib/certificate"
+)
+
+const (
+	commonName      = "kube-api-admission"
+	snapshotName    = "admission-webhook-client-key-pair-secrets"
+	secretName      = "admission-webhook-client-key-pair"
+	secretNamespace = "d8-system"
+)
+
+var _ = sdk.RegisterFunc(&go_hook.HookConfig{
+	OnBeforeHelm: &go_hook.OrderedConfig{Order: 10},
+	Kubernetes: []go_hook.KubernetesConfig{
+		{
+			Name:       snapshotName,
+			ApiVersion: "v1",
+			Kind:       "Secret",
+			NameSelector: &types.NameSelector{
+				MatchNames: []string{secretName},
+			},
+			NamespaceSelector: &types.NamespaceSelector{
+				NameSelector: &types.NameSelector{
+					MatchNames: []string{secretNamespace},
+				},
+			},
+			FilterFunc: filterAdmissionSecret,
+		},
+	},
+}, generateValidateWebhookCert)
+
+func filterAdmissionSecret(obj *unstructured.Unstructured) (go_hook.FilterResult, error) {
+	var secret v1.Secret
+	err := sdk.FromUnstructured(obj, &secret)
+	if err != nil {
+		return nil, err
+	}
+	return certificate.Certificate{
+		Cert: string(secret.Data["tls.crt"]),
+		Key:  string(secret.Data["tls.key"]),
+	}, nil
+}
+
+func generateValidateWebhookCert(input *go_hook.HookInput) error {
+	snapshots := input.Snapshots[snapshotName]
+
+	// Try reuse certificate from Secret
+	if len(snapshots) > 0 {
+		secret := snapshots[0].(certificate.Certificate)
```

Comment:
need rebase and use new snapshot mechanism https://github.com/deckhouse/deckhouse/pull/12513